src/util.c

```diff
@@ -493,10 +493,13 @@ static CK_RV token_login(P11PROV_CTX *provctx, CK_SLOT_ID slotid,
         if (retc != CKR_OK) {
             P11PROV_raise(provctx, retc, "Failed to close session %lu", s);
         }
-        return ret;
+        goto done;
     }
 
-    return p11prov_ctx_set_login_session(provctx, s);
+    ret = p11prov_ctx_set_login_session(provctx, s);
+done:
+    OPENSSL_cleanse(pin, pinlen);
```

Comment:
This will destroy the pin stored in the parsed_uri object or the main configuration context.
You should call this within the if (pw_cb) { } section only. Instead of in a common "done" at the end.

Reply:
Fixed.